storage/zynq7000-sdcard/sdcard.c

```diff
@@ -0,0 +1,1212 @@
+		host->dmaBuffer = NULL;
+		return -1;
+	}
+
+	if ((host->dmaBufferPhys & 3) != 0) {
+		munmap(host->dmaBuffer, SDCARD_MAX_TRANSFER);
+		host->dmaBuffer = NULL;
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static int sdhost_isr(unsigned int n, void *arg)
+{
+	/* SD Host controller interrupts are triggered by level, not by edge.
+	 * This means that ISR has to remove the reason for interrupt (zero out bit in STATUS register)
+	 * or disable the interrupt (zero out bit in SIGNAL_ENABLE register) - otherwise we get stuck
+	 * in infinite interrupt loop.
+	 */
+	sdcard_hostData_t *host = (sdcard_hostData_t *)arg;
+	uint32_t activeInterrupts = *(host->base + SDHOST_REG_INTR_STATUS);
+	*(host->base + SDHOST_REG_INTR_SIGNAL_ENABLE) &= ~activeInterrupts;
+	return 0;
+}
+
+
+static int _sdio_cmdExecutionWait(sdcard_hostData_t *host, uint32_t flags, time_t wait_us)
+{
+	int ret = -ETIME;
+	bool doResetCmd = false, doResetDat = false;
+
+	uint32_t flagsWithErrors = flags | SDHOST_ERROR_REASONS;
+	*(host->base + SDHOST_REG_INTR_SIGNAL_ENABLE) |= flagsWithErrors | SDHOST_INTR_CARD_OUT;
+	/* If there is a pending interrupt from flagsWithErrors set, it will fire now */
+	int waitret = condWait(host->eventCond, host->eventLock, wait_us);
+	if (waitret == -ETIME) {
+		/* It's a bit odd, but card ejection doesn't cause any errors - only timeouts
+		 * We may get a SDHOST_INTR_CARD_OUT, but only if physical card detect is available
+		 */
+		host->card.commandTimeouts++;
+	}
+
+	uint32_t val = *(host->base + SDHOST_REG_INTR_STATUS);
+	if ((val & SDHOST_ERROR_REASONS) != 0) {
+		doResetCmd = (val & SDHOST_INTR_CMD_ERRORS) != 0;
+		doResetDat = (val & SDHOST_INTR_DAT_ERRORS) != 0;
+		*(host->base + SDHOST_REG_INTR_STATUS) = SDHOST_ERROR_REASONS;
+		ret = -EIO;
+	}
+	else if ((val & flags) == flags) {
+		*(host->base + SDHOST_REG_INTR_STATUS) = flags;
+		ret = EOK;
+	}
+
+	if (doResetCmd || (ret == -ETIME)) {
+		sdhost_reset(host, CLOCK_CONTROL_RESET_CMD);
+	}
+
+	if (doResetDat || (ret == -ETIME)) {
+		sdhost_reset(host, CLOCK_CONTROL_RESET_DAT);
+	}
+
+	if ((val & SDHOST_INTR_CARD_OUT) != 0) {
+		condSignal(presenceEvents.cond);
+	}
+
+	*(host->base + SDHOST_REG_INTR_SIGNAL_ENABLE) &= ~flagsWithErrors;
+	return ret;
+}
+
+
+/* res is single u32 if isLongResponse == 0, array of 4 u32 otherwise */
+static int _sdio_cmdSend(sdcard_hostData_t *host, uint8_t cmd, uint32_t arg, uint32_t *res, uint16_t blockCount, int isLongResponse)
+{
+	sdhost_command_reg_t cmdFrame;
+	uint32_t val;
+
+	if (cmd >= MAX_SD_COMMANDS) {
+		return -EINVAL;
+	}
+
+	sdhost_command_data_t dataType = sdCmdMetadata[cmd].dataType;
+	if (dataType == CMD_INVALID) {
+		return -EINVAL;
+	}
+
+	val = *(host->base + SDHOST_REG_PRES_STATE) & PRES_STATE_BUSY_FLAGS;
+	if (val != 0) {
+		TRACE("busy %x", val);
+		return -EBUSY;
+	}
+
+	val = *(host->base + SDHOST_REG_INTR_STATUS) & (SDHOST_INTR_TRANSFER_DONE | SDHOST_INTR_CMD_DONE);
+	if (val != 0) {
+		TRACE("intr bits %x", val);
+		return -EBUSY;
+	}
```

Comment:
read and check can be performed inside of `if`

Reply:
I want to print which bits remain on for debug purposes, so I think it should stay like that